File: android/core/src/main/java/com/ianthetechie/core/SafeMultiply.kt
package com.ianthetechie.core

import uniffi.foobar.BinaryOperator
import uniffi.foobar.ComputationException

class SafeMultiply: BinaryOperator {
    override fun perform(lhs: Long, rhs: Long): Long {
        try {
            return Math.multiplyExact(lhs, rhs)
        } catch (e: ArithmeticException) {
            throw ComputationException.Overflow()
        }
    }
}

File: apple/Sources/Foobar/SafeMultiply.swift
import Foundation
import UniFFI

public class SafeMultiply: UniFFI.BinaryOperator {
    public func perform(lhs: Int64, rhs: Int64) throws -> Int64 {
        let (res, overflow) = lhs.multipliedReportingOverflow(by: rhs)

        if overflow {
            throw UniFFI.ComputationError.Overflow
        }

        return res
    }
}
